### ISSUE TITLE
Need advice for setting up testing environment.

### ISSUE THREAD

**Issue by booblebobble** [poliva/pyqtggpo#46]
I'm testing some additions I've written that involve both client and server modifications.  I've got a test server and test clients all running on my LAN (after editing handle_auth to let that happen), but actually trying to play matches between two such clients doesn't work.  The clients' emulators don't seem to be able to find each other and it stays stuck on the 'Connecting to Server' screen.

Is there something I can configure to let matches play locally or do I need to do something different entirely?  How do you usually handle testing?  My additions depend on monitoring the emulation as it's running to deliver live game information to the server, so it's important to be able to play actual games to be able to test thoroughly.

I've put preliminary versions of the client and server code on my github page if you'd like to see them.  In a nutshell, I added a matchmaking feature, a 'king of the hill' game based on the MKX gamemode, and a system for running automated tournaments, all built on a basic extension framework.

**Comment by poliva** [comment]
Hi and sorry for my delay in replaying, I've been out due to holidays those days!  First let me thank you for all the work you have done here, I've looked at it briefly and it looks awesome. Now let me try to answer your questions:

If you're using the ggpofba bundled into FightCade you'll need to change the hostname it connects to in the ggponet DLL: open the dll in an hex editor and replace ggpo-ng.com with the ip address of your test server (or ggpo.com if you're using the ggpofba from the original GGPO). Alternatively you can edit the hosts file on your test clients.

For my test environment I use a Mac with VMWare Fusion, and Linux and Windows virtual machines, this should be ok for most of your testing (for NAT traversal testing I use another laptop and a tethered 4G connection, but from what I have seen in your code you won't have to deal with NAT issues).

Let me know if you have any more questions, I will have more time and will be available for testing at the beginning of the year. Thanks again and happy holidays! :)

**Comment by booblebobble** [comment]
Yes I apologize for bothering you during the holidays.  We can talk more
next month.  I wouldn't mind a few days away from the project tbh lol.

I'd been using a hosts redirect of ggpo-ng.com and everything seemed to be
working with the exception of actually playing.  I'll try editing the dll
instead to see if that makes a difference or see if I can get some VMs set
up.

Thanks again, and happy holidays

Rob

**Comment by poliva** [comment]
I know i have this on my plate, just been damn busy with work and family :(.
If time permits it will start testing your code this weekend!

**Comment by booblebobble** [comment]
No problem at all, whenever you get the time.  I only hope it turns out to
be usable.

I wish I could do more on my end but for the life of me I can't get games
to run locally, even after reverting to the current unmodified server and
client.  However I arrange the server+clients on various machines I can't
get the server to talk to the FBA instances (Or rather, to *both* of them.
Usually one can communicate but not the other).  I've used a packet
analyzer to make sure nothing's still trying to hit the remote server.  I'm
not really much of a networking guy, so I'm a little hopeless with such
things :/  For the record, I can play on FC in the usual way with no
problem (after removing the hosts entry, of course)

I think the main things that I wasn't able to work on were
1) Verifying that it is correct to use the point where the automatic
spectator is added (Extension.MatchStarting) as a signal that the clients
are connected and the game is currently up and running for both of them.
2) That the current way to check that clients are out of a match and can be
paired up (ExtensionMatch._availability) is reliable.  In particular I'm
concerned that the quark (GGPOClient.quark) might not always get cleared if
a previous game fails to start.
3) The code to queue and add spectators to a spectating-restricted match
hasn't been tested at all, but it's fairly straightforward so hopefully it
doesn't cause much trouble.
4) Not too important, but currently to end a game I signal the clients to
kill the emulators themselves.  It might be more graceful if I could do it
locally, perhaps by manually ending the 'player' client.  ExtensionMatch
could be simplified slightly if I didn't have to wait for the match to be
cleaned up indirectly.

Also, I made a minor change to the monitoring code in the client so it
doesn't stop prematurely.  It's on my github.

Thanks,

Rob
